### PR TITLE
Add CT and UT to all documents and refactor

### PR DESCRIPTION
**Pull request by alexfca**
This is the 1st refactor for the Dendrite data which includes
- Create a "base" CosmosDocument that contains all the common properties
- Replace the explicit fields with the CosmosDocument
- Add Create Time to documents
- Add Update Time to documents
- Refactor

This is non-functional and DM + Room messages still work

For example here is a doc which has been updated (see _ct vs _ut)
```
    {
        "id": "matrix_roomserver_events,test.criticalarc.com,$Iq5D1mSuiyyu4uOXCMlqaIowbxRXNwXWq7DPXJfs0qQ",
        "_pk": "matrix_roomserver_events,test.criticalarc.com",
        "_sid": "test.criticalarc.com",
        "_cn": "matrix_roomserver_events",
        "_ct": "2021-09-20T05:33:53Z",
        "_ut": "2021-09-20T05:33:54Z",
        "_etag": "\"2b0065ab-0000-1a00-0000-61481d450000\"",
        "mx_roomserver_event": {
            "event_nid": 17,
            "room_nid": 3,
            "event_type_nid": 7,
            "event_state_key_nid": 1,
            "sent_to_output": true,
            "state_snapshot_nid": 30,
            "depth": 5,
            "event_id": "$Iq5D1mSuiyyu4uOXCMlqaIowbxRXNwXWq7DPXJfs0qQ",
            "reference_sha256": "Iq5D1mSuiyyu4uOXCMlqaIowbxRXNwXWq7DPXJfs0qQ=",
            "auth_event_nids": [
                13,
                15,
                14
            ],
            "is_rejected": false
        },
        "_rid": "bxhpANStt8KZAQAAAAAAAA==",
        "_self": "dbs/bxhpAA==/colls/bxhpANStt8I=/docs/bxhpANStt8KZAQAAAAAAAA==/",
        "_attachments": "attachments/",
        "_ts": 1632116037
    },
 
```

### CODE REVIEW

File: appservice/storage/cosmosdb/appservice_events_table.go [hunk 1]
```diff
@@ -57,13 +57,8 @@ type EventNumberCosmosData struct {
...
 
 type EventCosmosData struct {
-	Id        string      `json:"id"`
-	Pk        string      `json:"_pk"`
-	Tn        string      `json:"_sid"`
-	Cn        string      `json:"_cn"`
-	ETag      string      `json:"_etag"`
-	Timestamp int64       `json:"_ts"`
-	Event     EventCosmos `json:"mx_appservice_event"`
+	cosmosdbapi.CosmosDocument
```
Comment (alexfca) [inline appservice/storage/cosmosdb/appservice_events_table.go:60]: This is the example of the base class replacement

File: appservice/storage/cosmosdb/appservice_events_table.go [hunk 2]
```diff
@@ -166,6 +161,23 @@ func queryEventEventNumber(s *eventsStatements, ctx context.Context, qry string,
 	return response, nil
 }
 
+func getEvent(s *eventsStatements, ctx context.Context, pk string, docId string) (*EventCosmosData, error) {
```
Comment (alexfca) [inline appservice/storage/cosmosdb/appservice_events_table.go:164]: Added in some "getXXX()" methods where we do Upserts to allow us to set the UT for the existing doc

File: appservice/storage/cosmosdb/appservice_events_table.go [hunk 3]
```diff
@@ -338,46 +350,45 @@ func (s *eventsStatements) insertEvent(
...
-	// -1, // No transaction ID yet
-	data := EventCosmos{
-		AppServiceID:      appServiceID,
-		HeaderedEventJSON: eventJSON,
-		ID:                idSeq,
-		TXNID:             -1,
-	}
+	dbData, err := getEvent(s, ctx, pk, cosmosDocId)
+	if dbData != nil {
+		dbData.SetUpdateTime()
```
Comment (alexfca) [inline appservice/storage/cosmosdb/appservice_events_table.go:360]: Here is the example of setting the new time for existing docs

File: internal/cosmosdbapi/client.go [hunk 4]
```diff
@@ -22,3 +25,63 @@ func GetClient(conn CosmosConnection) *cosmosapi.Client {
 	}
 	return cosmosapi.New(conn.Url, cfg, nil, nil)
 }
+
+func UpsertDocument(ctx context.Context,
```
Comment (alexfca) [inline internal/cosmosdbapi/client.go:29]: New Upsert method. Will add others like this in upcoming PRs

File: internal/cosmosdbapi/document.go [hunk 5]
```diff
@@ -1,11 +1,21 @@
 package cosmosdbapi
 
 import (
-	"context"
 	"fmt"
 	"strings"
 )
 
+type CosmosDocument struct {
```
Comment (alexfca) [inline internal/cosmosdbapi/document.go:8]: This is the "common" base CosmosDocument

**Review by jaredjn — APPROVED** [review]
Wow, big chunk of work Alex. With my minimal go experience, LGTM

**Pull request merged**